zio-http/src/main/scala/zhttp/http/Http.scala

```diff
@@ -369,7 +369,31 @@ sealed trait Http[-R, +E, -A, +B] extends (A => ZIO[R, Option[E], B]) { self =>
    * should be stack safe. The performance improves quite significantly if no additional heap allocations are required
    * this way.
    */
-  final private[zhttp] def execute(a: A): HExit[R, E, B] =
+
+  final private[zhttp] def execute2[X](a: X)(implicit ev: HttpConverter[X, A]): HExit[R, E, B] = {
+    self match {
+      case Http.Empty         => HExit.empty
+      case Http.Identity      => HExit.succeed(ev.convert(a).asInstanceOf[B])
+      case Succeed(b)         => HExit.succeed(b)
+      case Fail(e)            => HExit.fail(e)
+      case FromFunctionZIO(f) => HExit.fromZIO(f(ev.convert(a)))
+      case Collect(pf)        => if (pf.isDefinedAt(ev.convert(a))) HExit.succeed(pf(ev.convert(a))) else HExit.empty
+      case Chain(self, other) => self.execute2(a).flatMap(b => other.execute2(b))
+      case Race(self, other)  =>
+        (self.execute2(a), other.execute2(a)) match {
+          case (HExit.Effect(self), HExit.Effect(other)) =>
+            Http.fromOptionFunction[Any](_ => self.raceFirst(other)).execute2(a)
+          case (HExit.Effect(_), other)                  => other
+          case (self, _)                                 => self
+        }
+
+      case FoldHttp(self, ee, bb, dd) =>
+        self.execute2(a).foldExit(ee(_).execute2(a), bb(_).execute2(a), dd.execute2(a))
+
+      case RunMiddleware(app, mid) => mid(app).execute2(a)
+    }
+  }
+  final private[zhttp] def execute(a: A): HExit[R, E, B]                                       =
```

Comment:
You can modify the current execute also. I don't think you need a fresh copy of the implementation